examples/reflection/dynamic_types.rs

```diff
@@ -0,0 +1,258 @@
+//! This example demonstrates the use of dynamic types in Bevy's reflection system.
+
+use bevy::reflect::{
+    reflect_trait, serde::TypedReflectDeserializer, std_traits::ReflectDefault, DynamicArray,
+    DynamicEnum, DynamicList, DynamicMap, DynamicStruct, DynamicTuple, DynamicTupleStruct,
+    DynamicVariant, FromReflect, Reflect, ReflectFromReflect, ReflectRef, TypeRegistry, Typed,
+};
+use serde::de::DeserializeSeed;
+use std::collections::HashMap;
+
+fn main() {
+    #[derive(Reflect, Default)]
+    #[reflect(Identifiable, Default)]
+    struct Player {
+        id: u32,
+    }
+
+    #[reflect_trait]
+    trait Identifiable {
+        fn id(&self) -> u32;
+    }
+
+    impl Identifiable for Player {
+        fn id(&self) -> u32 {
+            self.id
+        }
+    }
+
+    // Normally, when instantiating a type, you get back exactly that type.
+    // This is because the type is known at compile time.
+    // We call this the "concrete" or "canonical" type.
+    let player: Player = Player { id: 123 };
+
+    // When working with reflected types, however, we often "erase" this type information
+    // using the `Reflect` trait object.
+    // The underlying type is still the same (in this case, `Player`),
+    // but now we've hidden that information from the compiler.
+    let reflected: Box<dyn Reflect> = Box::new(player);
+
+    // Because it's the same type under the hood, we can still downcast it back to the original type.
+    assert!(reflected.downcast_ref::<Player>().is_some());
+
+    // But now let's "clone" our type using `Reflect::clone_value`.
+    let cloned: Box<dyn Reflect> = reflected.clone_value();
+
+    // If we try to downcast back to `Player`, we'll get an error.
+    assert!(cloned.downcast_ref::<Player>().is_none());
+
+    // Why is this?
+    // Well the reason is that `Reflect::clone_value` actually creates a dynamic type.
+    // Since `Player` is a struct, we actually get a `DynamicStruct` back.
+    assert!(cloned.is::<DynamicStruct>());
+
+    // This dynamic type is used to represent (or "proxy") the original type,
+    // so that we can continue to access its fields and overall structure.
+    let ReflectRef::Struct(cloned_ref) = cloned.reflect_ref() else {
+        panic!("expected struct")
+    };
+    let id = cloned_ref.field("id").unwrap().downcast_ref::<u32>();
+    assert_eq!(id, Some(&123));
+
+    // It also enables us to create a representation of a type without having compile-time
+    // access to the actual type. This is how the reflection deserializers work.
+    // They generally can't know how to construct a type ahead of time,
```

Comment:
Just so I understand; they could if we just provided an API `typed_deserialize<T> -> T`
(instead of the `deserialize -> dyn Reflect` that we have?

Reply:
Yep! I had considered mentioning how we _could_ automatically use `ReflectDefault` or `ReflectFromReflect` internally but I figured that was an unnecessary technical note